### PR TITLE
Use jni-rs to generate Java bindings. #68

### PR DESCRIPTION
**Pull request by dabreegster**
Proof of concept! What's the minimum needed to use this in https://github.com/BudgieInWA/joSM2Streets/? What's the simplest modern way to publish to Maven or similar?

### CODE REVIEW

File: osm2streets-java/StreetNetwork.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+import java.nio.file.Paths;
+
+class StreetNetwork {
+	static {
+		System.loadLibrary("osm2streets_java");
+	}
+
+	private static native long create(String osmXmlInput);
+
+	private static native String toGeojsonPlain(long pointer);
```
Comment (dabreegster) [inline osm2streets-java/StreetNetwork.java:11]: I gave up fighting to make these normal methods on a `StreetNetwork` object. The example code just passes are raw `long` pointers: https://github.com/jni-rs/jni-rs/blob/f8e78e3583ce850c2699c5e72cb8517d8fff09a6/example/HelloWorld.java#L7.
I think we can do this properly with `jobject`, but I'll leave that for a followup

File: osm2streets-java/run.sh [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+set -e
+javac StreetNetwork.java
+cargo build
+java -Djava.library.path=../target/x86_64-unknown-linux-gnu/debug/ StreetNetwork
```
Comment (dabreegster) [inline osm2streets-java/run.sh:6]: You probably need to change this on you end; it's likely just `../target/debug`. I have the architecture bit in there so compiling native and wasm32 doesn't stomp on the build directories

File: osm2streets-java/StreetNetwork.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+class StreetNetwork {
+	static {
+		System.loadLibrary("osm2streets_java");
```
Comment (dabreegster) [inline osm2streets-java/StreetNetwork.java:6]: You may need to prepend `lib` here. I had to fiddle around to get this to work. The actual file is `libosm2streets_java.so` for me

File: osm2streets-java/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                .unwrap();
+        let transformations = Transformation::standard_for_clipped_areas();
+        network.apply_transformations(transformations, &mut timer);
+
+        Self { inner: network }
+    }
+}
+
+#[no_mangle]
+pub extern "system" fn Java_StreetNetwork_create(
```
Comment (dabreegster) [inline osm2streets-java/src/lib.rs:29]: These signatures have to match the `.h`

File: osm2streets-java/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+#[no_mangle]
+pub unsafe extern "system" fn Java_StreetNetwork_toGeojsonPlain(
+    env: JNIEnv,
+    _: JClass,
+    pointer: jlong,
+) -> jstring {
+    let streets = &mut *(pointer as *mut StreetNetwork);
+    let result = streets.inner.to_geojson().unwrap();
```
Comment (dabreegster) [inline osm2streets-java/src/lib.rs:46]: Once we figure out how to use a normal Java object, we can hopefully just call normal methods in the `impl` block above

**Comment by BudgieInWA** [comment]
This runs for me, after tweaking the commands.

I don't know at all about packaging/publishing on maven, or packaging the built binary. I have to reference the class in the plugin build, which is currently targeting java 11 for some reason.

I'd be keen to jump on a call to try to figure it out some time.

**Comment by dabreegster** [comment]
I'll start reading about the packaging process. And a call would be good; it's maybe also time to consider putting the WASM+JS module on NPM and talking about versioning. Next week is pretty open on my end so far

**Comment by dabreegster** [comment]
Thanks to help from https://github.com/jni-rs/jni-rs/issues/394, the Java API is now sensible. If there's no opposition, I'll merge this tomorrow and look into Maven.

**Review by BudgieInWA — APPROVED** [review]
Looks good!

I was thinking of trying to manually include the dependency (via the .class file or something), but that means figuring out how to get the lib portion in there. Looking into publishing for Maven might actually be the best place to find the answers to those problems.

**Pull request merged**